File: Data Structures & Algorithms/balanced-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        def height(root)->int:
            if not root:
                return 0
            return max(height(root.left),height(root.right))+1
        l=height(root.left)
        r=height(root.right)
        if abs(l-r)<=1 and self.isBalanced(root.left) and self.isBalanced(root.right):
            return True
        return False
        

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int ind=0;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n=preorder.length;
        return solve(preorder,inorder,0,n-1);

    }
    TreeNode solve(int[] pre,int[] in,int st,int end){
        if(st>end)return null;
        int rootval=pre[ind++];
        int i=st;
        for(;i<=end;i++){
            if(in[i]==rootval)break;
        }
       
        TreeNode root=new TreeNode(rootval);
        root.left=solve(pre,in,st,i-1);
        root.right=solve(pre,in,i+1,end);
        return root;
    }
}
